### PR TITLE
feat: add event tracing

### PR DESCRIPTION
**Pull request by hperl**
This implements the following events from https://github.com/ory-corp/cloud/issues/4028:

OAuth2 & OIDC

- OIDC initiated login succeeded
- OIDC initiated login failed
- Consent granted & revoked
- Consent grant aborted

Machine activity: Tokens, Permission checks
OAuth2

- Access tokens created
- Access token not created
- Access token inspected
- Access token check with result

## Checklist

- [ ] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md).
- [ ] I have referenced an issue containing the design document if my change
      introduces a new feature.
- [ ] I am following the
      [contributing code guidelines](../blob/master/CONTRIBUTING.md#contributing-code).
- [ ] I have read the [security policy](../security/policy).
- [ ] I confirm that this pull request does not address a security
      vulnerability. If this pull request addresses a security vulnerability, I
      confirm that I got the approval (please contact
      [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push
      the changes.
- [ ] I have added tests that prove my fix is effective or that my feature
      works.
- [ ] I have added or changed [the documentation](https://github.com/ory/docs).

### CODE REVIEW

**Review by kmherrmann — APPROVED** [review]
Content-wise looks great, although i can't vouch for all code changes :)

File: x/events/events.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+	ConsentRejected semconv.Event = "OAuth2ConsentRejected"
+
+	// ConsentRevoked will be emitted when the user revokes a consent request.
+	ConsentRevoked semconv.Event = "OAuth2ConsentRevoked"
+
+	// AccessTokenCreated will be emitted by requests to POST /oauth2/token in case the request was successful.
+	AccessTokenCreated semconv.Event = "OAuth2AccessTokenCreated" //nolint:gosec
+
+	// AccessTokenNotCreated will be emitted by requests to POST /oauth2/token in case the request was unsuccessful.
+	AccessTokenNotCreated semconv.Event = "OAuth2AccessTokenNotCreated" //nolint:gosec
```
Comment (alnr) [inline x/events/events.go:34]: This feels weird. Do we really need this?

Comment (hperl) [reply x/events/events.go:34]: It was in the list of needed events, so I guess? @kmherrmann ?

File: oauth2/handler.go [hunk 2]
```diff
@@ -885,16 +893,19 @@ func (h *Handler) oauth2TokenExchange(w http.ResponseWriter, r *http.Request) {
 	if err != nil {
 		h.logOrAudit(err, r)
 		h.r.OAuth2Provider().WriteAccessError(ctx, w, accessRequest, err)
+		events.Trace(ctx, events.AccessTokenNotCreated)
```
Comment (aeneasr) [inline oauth2/handler.go:896]: Maybe OAuth2TokenExchangeError?

File: x/events/events.go (hunk 1 above)
Comment (aeneasr) [reply x/events/events.go:36]: Maybe make it more generic and call it OAuh2FlowError ?

File: oauth2/handler.go (hunk 2 above)
Comment (hperl) [reply oauth2/handler.go]: Fixed, thanks!

**Comment by aeneasr** [comment]
Panic: https://github.com/ory/hydra/actions/runs/5308371608/jobs/9607810461#step:6:223

**Comment by hperl** [comment]
> Panic: https://github.com/ory/hydra/actions/runs/5308371608/jobs/9607810461#step:6:223

fixed :)

File: x/events/events.go (hunk 1 above)
Comment (alnr) [reply x/events/events.go]: We probably don't need this event at all. What it's trying to signal is some error during the OAuth2 flow. But many many things can go wrong during the flow, and we're not about to signal all of them to our analytics platform, are we?

Errors typically don't get emitted as analytics events but go to the monitoring platform (Grafana) as logs or error traces.

Comment (aeneasr) [reply x/events/events.go]: Good points Arne! I think we can still ship this for now and invest some time in better observability next. We don't have to use the events and if they serve no purpose in Analytics, we'll remove them

**Pull request merged**